// ngx-useful-swiper/bundles/ngx-useful-swiper.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('swiper/bundle')) :
    typeof define === 'function' && define.amd ? define('ngx-useful-swiper', ['exports', '@angular/common', '@angular/core', 'swiper/bundle'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-useful-swiper'] = {}, global.ng.common, global.ng.core, global.Swiper));
}(this, (function (exports, common, core, Swiper) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var Swiper__default = /*#__PURE__*/_interopDefaultLegacy(Swiper);

    var SwiperComponent = /** @class */ (function () {
        function SwiperComponent(elementRef, changeDetectorRef) {
            this.elementRef = elementRef;
            this.changeDetectorRef = changeDetectorRef;
            this.slideCount = 0;
            this.initialized = false;
            this.shouldInitialize = true;
        }
        Object.defineProperty(SwiperComponent.prototype, "initialize", {
            set: function (value) {
                this.shouldInitialize = this.initialized ? false : value;
            },
            enumerable: false,
            configurable: true
        });
        SwiperComponent.prototype.ngAfterViewInit = function () {
            if (this.shouldInitialize) {
                this.setup();
            }
        };
        SwiperComponent.prototype.setup = function () {
            if (!this.swiper) {
                // if rendered on server querySelector is undefined
                if (this.elementRef.nativeElement.querySelector) {
                    this.swiperWrapper = this.elementRef.nativeElement.querySelector('.swiper-wrapper');
                    this.slideCount = this.swiperWrapper.childElementCount;
                    this.swiper = new Swiper__default['default'](this.elementRef.nativeElement.querySelector('swiper > div'), this.config);
                    this.changeDetectorRef.detectChanges();
                }
                this.shouldInitialize = false;
            }
        };
        SwiperComponent.prototype.ngAfterViewChecked = function () {
            if (this.shouldInitialize) {
                this.setup();
            }
            if (this.swiperWrapper &&
                this.slideCount !== this.swiperWrapper.childElementCount) {
                this.slideCount = this.swiperWrapper.childElementCount;
                this.swiper.update();
            }
        };
        SwiperComponent.prototype.ngOnDestroy = function () {
            this.swiper.destroy(true, true);
        };
        return SwiperComponent;
    }());
    SwiperComponent.decorators = [
        { type: core.Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: 'swiper',
                    template: "<div\r\n  [ngClass]=\"{\r\n    'swiper-container': config?.containerModifierClass === undefined\r\n  }\"\r\n>\r\n  <ng-content></ng-content>\r\n</div>\r\n",
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    styles: [":host{display:block}.swiper-container{height:100%;width:100%}"]
                },] }
    ];
    SwiperComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };
    SwiperComponent.propDecorators = {
        config: [{ type: core.Input }],
        initialize: [{ type: core.Input, args: ['initialize',] }]
    };

    var NgxUsefulSwiperModule = /** @class */ (function () {
        function NgxUsefulSwiperModule() {
        }
        return NgxUsefulSwiperModule;
    }());
    NgxUsefulSwiperModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [SwiperComponent],
                    imports: [common.CommonModule],
                    exports: [SwiperComponent]
                },] }
    ];

    /*
     * Public API Surface of ngx-useful-swiper
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxUsefulSwiperModule = NgxUsefulSwiperModule;
    exports.SwiperComponent = SwiperComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
